# Gather 


Gathering Data for this Project composed from three pieces of data as described below:
- The WeRateDogs Twitter archive. We will download this file manually by clicking the following link: twitter_archive_enhanced.csv 
- The tweet image predictions, i.e., what breed of dog (or other object, animal, etc.) is present in each tweet according to a neural network. This file (image_predictions.tsv) hosted on Udacity's servers and should be downloaded programmatically using the Requests library and the following URL: https://d17h27t6h515a5.cloudfront.net/topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv 
- Each tweet's retweet count and favorite (i.e. "like") count at minimum, and any additional data we will find interesting. Using the tweet IDs in the WeRateDogs Twitter archive, we will query the Twitter API for each tweet's JSON data using Python's Tweepy library and store each tweet's entire set of JSON data in a file called tweet_json.txt file. Each tweet's JSON data should be written to its own line. Then we will read this .txt file line by line into a pandas DataFrame with (at minimum) tweet ID, retweet count, and favorite count.


In [208]:
# Import the libraries that we will need in this project
import pandas as pd
import requests
import numpy
import tweepy
import json
import os
import time
import re

In [66]:
# Read the twitter-archive-enhanced.csv file and store it as dataframe in archive
archive = pd.read_csv('twitter-archive-enhanced.csv')


tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0     892420643555336193                    NaN                  NaN   
1     892177421306343426                    NaN                  NaN   
2     891815181378084864                    NaN                  NaN   
3     891689557279858688                    NaN                  NaN   
4     891327558926688256                    NaN                  NaN   
5     891087950875897856                    NaN                  NaN   
6     890971913173991426                    NaN                  NaN   
7     890729181411237888                    NaN                  NaN   
8     890609185150312448                    NaN                  NaN   
9     890240255349198849                    NaN                  NaN   
10    890006608113172480                    NaN                  NaN   
11    889880896479866881                    NaN                  NaN   
12    889665388333682689                    NaN                  NaN   
13    889638837579907072                    NaN                  NaN   
14    889531135344209921                    NaN                  NaN   
15    889278841981685760                    NaN                  NaN   
16    888917238123831296                    NaN                  NaN   
17    888804989199671297                    NaN                  NaN   
18    888554962724278272                    NaN                  NaN   
19    888202515573088257                    NaN                  NaN   
20    888078434458587136                    NaN                  NaN   
21    887705289381826560                    NaN                  NaN   
22    887517139158093824                    NaN                  NaN   
23    887473957103951883                    NaN                  NaN   
24    887343217045368832                    NaN                  NaN   
25    887101392804085760                    NaN                  NaN   
26    886983233522544640                    NaN                  NaN   
27    886736880519319552                    NaN                  NaN   
28    886680336477933568                    NaN                  NaN   
29    886366144734445568                    NaN                  NaN   
...                  ...                    ...                  ...   
2326  666411507551481857                    NaN                  NaN   
2327  666407126856765440                    NaN                  NaN   
2328  666396247373291520                    NaN                  NaN   
2329  666373753744588802                    NaN                  NaN   
2330  666362758909284353                    NaN                  NaN   
2331  666353288456101888                    NaN                  NaN   
2332  666345417576210432                    NaN                  NaN   
2333  666337882303524864                    NaN                  NaN   
2334  666293911632134144                    NaN                  NaN   
2335  666287406224695296                    NaN                  NaN   
2336  666273097616637952                    NaN                  NaN   
2337  666268910803644416                    NaN                  NaN   
2338  666104133288665088                    NaN                  NaN   
2339  666102155909144576                    NaN                  NaN   
2340  666099513787052032                    NaN                  NaN   
2341  666094000022159362                    NaN                  NaN   
2342  666082916733198337                    NaN                  NaN   
2343  666073100786774016                    NaN                  NaN   
2344  666071193221509120                    NaN                  NaN   
2345  666063827256086533                    NaN                  NaN   
2346  666058600524156928                    NaN                  NaN   
2347  666057090499244032                    NaN                  NaN   
2348  666055525042405380                    NaN                  NaN   
2349  666051853826850816                    NaN         

In [67]:
# Read the image predictions file from the url link and store it tsv file
url = 'https://d17h27t6h515a5.cloudfront.net/topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv'
response = requests.get(url)
file_name = url.split('/')[-1]
with open(os.path.join(file_name), mode = 'wb') as file:
    file.write(response.content)

In [106]:
# Read the image predictions tsv file in pandas dataframe
image_predictions = pd.read_csv('image-predictions.tsv', sep = '\t')
image_predictions

tweet_id                                            jpg_url  \
0     666020888022790149    https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1     666029285002620928    https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2     666033412701032449    https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3     666044226329800704    https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4     666049248165822465    https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   
5     666050758794694657    https://pbs.twimg.com/media/CT5Jof1WUAEuVxN.jpg   
6     666051853826850816    https://pbs.twimg.com/media/CT5KoJ1WoAAJash.jpg   
7     666055525042405380    https://pbs.twimg.com/media/CT5N9tpXIAAifs1.jpg   
8     666057090499244032    https://pbs.twimg.com/media/CT5PY90WoAAQGLo.jpg   
9     666058600524156928    https://pbs.twimg.com/media/CT5Qw94XAAA_2dP.jpg   
10    666063827256086533    https://pbs.twimg.com/media/CT5Vg_wXIAAXfnj.jpg   
11    666071193221509120    https://pbs.twimg.com/media/CT5cN_3WEAAlOoZ.jpg   
12    666073100786774016    https://pbs.twimg.com/media/CT5d9DZXAAALcwe.jpg   
13    666082916733198337    https://pbs.twimg.com/media/CT5m4VGWEAAtKc8.jpg   
14    666094000022159362    https://pbs.twimg.com/media/CT5w9gUW4AAsBNN.jpg   
15    666099513787052032    https://pbs.twimg.com/media/CT51-JJUEAA6hV8.jpg   
16    666102155909144576    https://pbs.twimg.com/media/CT54YGiWUAEZnoK.jpg   
17    666104133288665088    https://pbs.twimg.com/media/CT56LSZWoAAlJj2.jpg   
18    666268910803644416    https://pbs.twimg.com/media/CT8QCd1WEAADXws.jpg   
19    666273097616637952    https://pbs.twimg.com/media/CT8T1mtUwAA3aqm.jpg   
20    666287406224695296    https://pbs.twimg.com/media/CT8g3BpUEAAuFjg.jpg   
21    666293911632134144    https://pbs.twimg.com/media/CT8mx7KW4AEQu8N.jpg   
22    666337882303524864    https://pbs.twimg.com/media/CT9OwFIWEAMuRje.jpg   
23    666345417576210432    https://pbs.twimg.com/media/CT9Vn7PWoAA_ZCM.jpg   
24    666353288456101888    https://pbs.twimg.com/media/CT9cx0tUEAAhNN_.jpg   
25    666362758909284353    https://pbs.twimg.com/media/CT9lXGsUcAAyUFt.jpg   
26    666373753744588802    https://pbs.twimg.com/media/CT9vZEYWUAAlZ05.jpg   
27    666396247373291520    https://pbs.twimg.com/media/CT-D2ZHWIAA3gK1.jpg   
28    666407126856765440    https://pbs.twimg.com/media/CT-NvwmW4AAugGZ.jpg   
29    666411507551481857    https://pbs.twimg.com/media/CT-RugiWIAELEaq.jpg   
...                  ...                                                ...   
2045  886366144734445568    https://pbs.twimg.com/media/DE0BTnQUwAApKEH.jpg   
2046  886680336477933568    https://pbs.twimg.com/media/DE4fEDzWAAAyHMM.jpg   
2047  886736880519319552    https://pbs.twimg.com/media/DE5Se8FXcAAJFx4.jpg   
2048  886983233522544640    https://pbs.twimg.com/media/DE8yicJW0AAAvBJ.jpg   
2049  887101392804085760    https://pbs.twimg.com/media/DE-eAq6UwAA-jaE.jpg   
2050  887343217045368832  https://pbs.twimg.com/ext_tw_video_thumb/88734...   
2051  887473957103951883    https://pbs.twimg.com/media/DFDw2tyUQAAAFke.jpg   
2052  887517139158093824  https://pbs.twimg.com/ext_tw_video_thumb/88751...   
2053  887705289381826560    https://pbs.twimg.com/media/DFHDQBbXgAEqY7t.jpg   
2054  888078434458587136    https://pbs.twimg.com/media/DFMWn56WsAAkA7B.jpg   
2055  888202515573088257    https://pbs.twimg.com/media/DFDw2tyUQAAAFke.jpg   
2056  888554962724278272    https://pbs.twimg.com/media/DFTH_O-UQAACu20.jpg   
2057  888804989199671297    https://pbs.twimg.com/media/DFWra-3VYAA2piG.jpg   
2058  888917238123831296    https://pbs.twimg.com/media/DFYRgsOUQAARGhO.jpg   
2059  889278841981685760  https://pbs.twimg.com/ext_tw_video_thumb/88927...   
2060  889531135344209921    https://pbs.twimg.com/media/DFg_2PVW0AEHN3p.jpg   
2061  889638837579907072    https://pbs.twimg.com/media/DFihzFfXsAYGDPR.jpg   
2062  889665388333682689    https://pbs.twimg.com/media/DFi579UWsAAatzw.jpg   
2063  889880896479866881    https://pbs.twimg.com/media/DFl99B1WsA

In [69]:
# auth information for twitter api
consumer_key = 'eb62oIuNrHdMUfKZCQ0SaDUPO'
consumer_secret = 'k7GVwH6hOp2WNgBnivl0hsg806WtFp3mSajHWrlrUEx1yZXmjd'
access_token = '3388506089-1hSZiRrOE0Jet8Wca4J1zoT0GjwqVpoJxVfQikP'
access_secret = 'dqlr9D5Kw02H4bG1cmsdpInKb9tsSjghRdvNHCI7253D1'

auth = tweepy.OAuthHandler(consumer_key, consumer_secret)
auth.set_access_token(access_token, access_secret)

# Construct the API instance
api = tweepy.API(auth, 
                 parser = tweepy.parsers.JSONParser(), # Parse the result to Json Object
                 wait_on_rate_limit = True, # Automatically wait for rate limits to replenish
                 wait_on_rate_limit_notify = True) # Print a notification when Tweepy is waiting for rate limits to replenish

In [75]:
# List where we will store the dictionaries of our result
df_tweet = []
# List frame where we will store the tweet_id of the errors
error_list = []


# Calculate the time of excution
start = time.time()

for tweetid in archive['tweet_id']:
    try:
        tweet = api.get_status(tweetid,tweet_mode = 'extended')
        # Print one page to look at the structure of the returned file
        # and the names of attributes
        #print(json.dumps(tweet, indent = 4))
        
        
        favorites = tweet['favorite_count'] # How many favorites the tweet had
        retweets = tweet['retweet_count'] # Count of the retweet
        user_followers = tweet['user']['followers_count'] # How many followers the user had
        user_favourites = tweet['user']['favourites_count'] # How many favorites the user had
        date_time = tweet['created_at'] # The date and time of the creation
        
        df_tweet.append({'tweet_id':int(tweetid),
                        'favorite_count':int(favorites),
                        'retweet_count':int(retweets),
                        'followers_count':int(user_followers),
                        'favourites_count':int(user_favourites),
                        'created_at':pd.to_datetime(date_time)}                       
        )
        #print (df_tweet)


        
        
    except Exception as e:
        print (str(tweetid) + " :"+ str(e) )
        error_list.append(tweetid)

# Calculate the time of excution       
end = time.time()
print(end - start)


888202515573088257 :[{'message': 'No status found with that ID.', 'code': 144}]
873697596434513921 :[{'message': 'No status found with that ID.', 'code': 144}]
869988702071779329 :[{'message': 'No status found with that ID.', 'code': 144}]
866816280283807744 :[{'message': 'No status found with that ID.', 'code': 144}]
861769973181624320 :[{'message': 'No status found with that ID.', 'code': 144}]
845459076796616705 :[{'message': 'No status found with that ID.', 'code': 144}]
842892208864923648 :[{'message': 'No status found with that ID.', 'code': 144}]
837012587749474308 :[{'message': 'No status found with that ID.', 'code': 144}]
827228250799742977 :[{'message': 'No status found with that ID.', 'code': 144}]
802247111496568832 :[{'message': 'No status found with that ID.', 'code': 144}]
775096608509886464 :[{'message': 'No status found with that ID.', 'code': 144}]


Rate limit reached. Sleeping for: 397


1686.3830001354218


In [79]:
# lengh of the result
print("The lengh of the result", len(df_tweet))
# The tweet_id of the errors
print("The lengh of the errors", len(error_list))

The lengh of the result 2345
The lengh of the errors 11


From the above results:
- We reached the limit of the tweepy API once but wait_on_rate_limit automatically wait for rate limits to replenish and wait_on_rate_limit_notify print a notification when Tweepy is waiting
- The total time was about 1686 seconds (~ 28 min)
- We could get 2345 tweet_id correctly with 11 errors (we will query those 7 errors separately)


In [82]:
# We repeat the same operation for the tweet_ids that we coudln't get and append the result to df_tweet
err_lst=[]
for tweetid in error_list:
    try:
        #tweet = api.get_status(tweetid,tweet_mode = 'extended')
        # Print one page to look at the structure of the returned file
        # and the names of attributes
        #print(json.dumps(tweet, indent = 4))
        
        
        favorites = tweet['favorite_count'] # How many favorites the tweet had
        retweets = tweet['retweet_count'] # Count of the retweet
        user_followers = tweet['user']['followers_count'] # How many followers the user had
        user_favourites = tweet['user']['favourites_count'] # How many favorites the user had
        date_time = tweet['created_at'] # The date and time of the creation
        
        df_tweet.append({'tweet_id':int(tweetid),
                        'favorite_count':int(favorites),
                        'retweet_count':int(retweets),
                        'followers_count':int(user_followers),
                        'favourites_count':int(user_favourites),
                        'created_at':pd.to_datetime(date_time)}                       
        )
        #print (df_tweet)


        
        
    except Exception as e:
        print (str(tweetid) + " :"+ str(e) )
        err_lst.append(tweetid)


In [84]:
print("The lengh of the result after Querying the errors separately", len(df_tweet))

The lengh of the result after Querying the errors separately 2356


In [89]:
# Create DataFrames from list of dictionaries
json_tweets = pd.DataFrame(df_tweet)

# Save the dataFrame in file
json_tweets.to_csv('tweet_json.txt', encoding = 'utf-8', index=False)


In [90]:
# Read the saved tweet_json.txt file into a dataframe
json_tweets = pd.read_csv('tweet_json.txt', encoding = 'utf-8')


created_at  favorite_count  favourites_count  followers_count  \
0     2017-08-01 16:23:56           39014            132501          6197725   
1     2017-08-01 00:17:27           33403            132501          6197725   
2     2017-07-31 00:18:03           25189            132501          6197725   
3     2017-07-30 15:58:51           42368            132501          6197725   
4     2017-07-29 16:00:24           40523            132501          6197725   
5     2017-07-29 00:08:17           20327            132501          6197726   
6     2017-07-28 16:27:12           11917            132501          6197726   
7     2017-07-28 00:22:40           65932            132501          6197726   
8     2017-07-27 16:25:51           27911            132501          6197726   
9     2017-07-26 15:59:51           32118            132501          6197726   
10    2017-07-26 00:31:25           30814            132501          6197726   
11    2017-07-25 16:11:53           27932            132501          6197726   
12    2017-07-25 01:55:32           48332            132501          6197727   
13    2017-07-25 00:10:02           27341            132501          6197727   
14    2017-07-24 17:02:04           15174            132501          6197728   
15    2017-07-24 00:19:32           25430            132501          6197728   
16    2017-07-23 00:22:39           29252            132501          6197728   
17    2017-07-22 16:56:37           25740            132501          6197728   
18    2017-07-22 00:23:06           20036            132501          6197728   
19    2017-07-20 16:49:33           21886            132501          6197728   
20    2017-07-19 16:06:48           30358            132501          6197729   
21    2017-07-19 03:39:09           46474            132501          6197729   
22    2017-07-19 00:47:34           69510            132501          6197729   
23    2017-07-18 16:08:03           33857            132501          6197730   
24    2017-07-18 00:07:08           30722            132501          6197731   
25    2017-07-17 16:17:36           35380            132501          6197731   
26    2017-07-16 23:58:41           12147            132501          6197731   
27    2017-07-16 20:14:00           22560            132501          6197731   
28    2017-07-15 23:25:31           21241            132501          6197732   
29    2017-07-15 16:51:35             117            132501          6197733   
...                   ...             ...               ...              ...   
2326  2015-11-16 14:57:41             105            132501          6198020   
2327  2015-11-16 04:02:55           14496            132501          6198020   
2328  2015-11-16 03:55:04              80            132501          6198019   
2329  2015-11-16 03:44:34             158            132501          6198020   
2330  2015-11-16 03:22:39             166            132501          6198020   
2331  2015-11-16 02:38:37             119            132501          6198019   
2332  2015-11-16 01:59:36             326            132501          6198021   
2333  2015-11-16 01:52:02             149            132501          6198021   
2334  2015-11-16 01:22:45             482            132501          6198021   
2335  2015-11-16 01:01:59             114            132501          6198021   
2336  2015-11-16 00:55:59             299            132501          6198021   
2337  2015-11-16 00:49:46             441            132501          6198021   
2338  2015-11-16 00:35:11            1236            132501          6198021   
2339  2015-11-16 00:30:50             133            132501          6198021   
2340  2015-11-16 00:24:50             109            132501          6198021   
2341  2015-11-16 00:04:52             302            132501          6198022   
2342  2015-11-15 23:21:54             125            132501          6198022   
2343  2015-11-15 23:05:30             131            132501          6198023   
2344  2015-11-1

After asessing the json dataset downloaded from twitter , i decided to remove the user's follower count and favorite count because those attributes , at a glance are not providing suplemental information. Perhaps that is the area for Exploratory data analysis to check if the said attributes add any value to the analysis

In [119]:
json_tweets = json_tweets.drop(['favourites_count','followers_count'], axis = 1)


# Gather: Summary 

Gathering is the first step in the data wrangling process. We could finish the high-level gathering process:

- Getting data from an existing file (twitter-archive-enhanced.csv) Reading from csv file using pandas 
- Downloading a file from the internet (image-predictions.tsv) Downloading file using requests 
- Querying an API (tweet_json.txt) Get JSON object of all the tweet_ids using Tweepy 





# Assess Data

After gathering each of the above pieces of data we will assess data visually and programmatically for quality and tidiness issues. Quality issues and tidiness issues will be documented in wrangle_act.ipynb Jupyter Notebook. 

In [97]:
# View first 20 rows of twitter_archive DataFrame
archive.head(20)


tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0   892420643555336193                    NaN                  NaN   
1   892177421306343426                    NaN                  NaN   
2   891815181378084864                    NaN                  NaN   
3   891689557279858688                    NaN                  NaN   
4   891327558926688256                    NaN                  NaN   
5   891087950875897856                    NaN                  NaN   
6   890971913173991426                    NaN                  NaN   
7   890729181411237888                    NaN                  NaN   
8   890609185150312448                    NaN                  NaN   
9   890240255349198849                    NaN                  NaN   
10  890006608113172480                    NaN                  NaN   
11  889880896479866881                    NaN                  NaN   
12  889665388333682689                    NaN                  NaN   
13  889638837579907072                    NaN                  NaN   
14  889531135344209921                    NaN                  NaN   
15  889278841981685760                    NaN                  NaN   
16  888917238123831296                    NaN                  NaN   
17  888804989199671297                    NaN                  NaN   
18  888554962724278272                    NaN                  NaN   
19  888202515573088257                    NaN                  NaN   

                    timestamp  \
0   2017-08-01 16:23:56 +0000   
1   2017-08-01 00:17:27 +0000   
2   2017-07-31 00:18:03 +0000   
3   2017-07-30 15:58:51 +0000   
4   2017-07-29 16:00:24 +0000   
5   2017-07-29 00:08:17 +0000   
6   2017-07-28 16:27:12 +0000   
7   2017-07-28 00:22:40 +0000   
8   2017-07-27 16:25:51 +0000   
9   2017-07-26 15:59:51 +0000   
10  2017-07-26 00:31:25 +0000   
11  2017-07-25 16:11:53 +0000   
12  2017-07-25 01:55:32 +0000   
13  2017-07-25 00:10:02 +0000   
14  2017-07-24 17:02:04 +0000   
15  2017-07-24 00:19:32 +0000   
16  2017-07-23 00:22:39 +0000   
17  2017-07-22 16:56:37 +0000   
18  2017-07-22 00:23:06 +0000   
19  2017-07-21 01:02:36 +0000   

                                               source  \
0   <a href="http://twitter.com/download/iphone" r...   
1   <a href="http://twitter.com/download/iphone" r...   
2   <a href="http://twitter.com/download/iphone" r...   
3   <a href="http://twitter.com/download/iphone" r...   
4   <a href="http://twitter.com/download/iphone" r...   
5   <a href="http://twitter.com/download/iphone" r...   
6   <a href="http://twitter.com/download/iphone" r...   
7   <a href="http://twitter.com/download/iphone" r...   
8   <a href="http://twitter.com/download/iphone" r...   
9   <a href="http://twitter.com/download/iphone" r...   
10  <a href="http://twitter.com/download/iphone" r...   
11  <a href="http://twitter.com/download/iphone" r...   
12  <a href="http://twitter.com/download/iphone" r...   
13  <a href="http://twitter.com/download/iphone" r...   
14  <a href="http://twitter.com/download/iphone" r...   
15  <a href="http://twitter.com/download/iphone" r...   
16  <a href="http://twitter.com/download/iphone" r...   
17  <a href="http://twitter.com/download/iphone" r...   
18  <a href="http://twitter.com/download/iphone" r...   
19  <a href="http://twitter.com/download/iphone" r...   

                                                 text  retweeted_status_id  \
0   This is Phineas. He's a mystical boy. Only eve...                  NaN   
1   This is Tilly. She's just checking pup on you....                  NaN   
2   This is Archie. He is a rare Norwegian Pouncin...                  NaN   
3   This is Darla. She commenced a snooze mid meal...                  NaN   
4   This is Franklin. He would like you to stop ca...                  NaN   
5   Here we have a majestic great white breaching ...                  NaN   
6   Meet Jax. He enjoys ice cream so much he gets ...                  NaN   
7   When you watch your own

In [98]:
archive.tail()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
2351  666049248165822465                    NaN                  NaN   
2352  666044226329800704                    NaN                  NaN   
2353  666033412701032449                    NaN                  NaN   
2354  666029285002620928                    NaN                  NaN   
2355  666020888022790149                    NaN                  NaN   

                      timestamp  \
2351  2015-11-16 00:24:50 +0000   
2352  2015-11-16 00:04:52 +0000   
2353  2015-11-15 23:21:54 +0000   
2354  2015-11-15 23:05:30 +0000   
2355  2015-11-15 22:32:08 +0000   

                                                 source  \
2351  <a href="http://twitter.com/download/iphone" r...   
2352  <a href="http://twitter.com/download/iphone" r...   
2353  <a href="http://twitter.com/download/iphone" r...   
2354  <a href="http://twitter.com/download/iphone" r...   
2355  <a href="http://twitter.com/download/iphone" r...   

                                                   text  retweeted_status_id  \
2351  Here we have a 1949 1st generation vulpix. Enj...                  NaN   
2352  This is a purebred Piers Morgan. Loves to Netf...                  NaN   
2353  Here is a very happy pup. Big fan of well-main...                  NaN   
2354  This is a western brown Mitsubishi terrier. Up...                  NaN   
2355  Here we have a Japanese Irish Setter. Lost eye...                  NaN   

      retweeted_status_user_id retweeted_status_timestamp  \
2351                       NaN                        NaN   
2352                       NaN                        NaN   
2353                       NaN                        NaN   
2354                       NaN                        NaN   
2355                       NaN                        NaN   

                                          expanded_urls  rating_numerator  \
2351  https://twitter.com/dog_rates/status/666049248...                 5   
2352  https://twitter.com/dog_rates/status/666044226...                 6   
2353  https://twitter.com/dog_rates/status/666033412...                 9   
2354  https://twitter.com/dog_rates/status/666029285...                 7   
2355  https://twitter.com/dog_rates/status/666020888...                 8   

      rating_denominator  name doggo floofer pupper puppo  
2351                  10  None  None    None   None  None  
2352                  10     a  None    None   None  None  
2353                  10     a  None    None   None  None  
2354                  10     a  None    None   None  None  
2355                  10  None  None    None   None  None

In [101]:
archive.info()
archive['name'].value_counts()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2356 non-null int64
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2356 non-null object
source                        2356 non-null object
text                          2356 non-null object
retweeted_status_id           181 non-null float64
retweeted_status_user_id      181 non-null float64
retweeted_status_timestamp    181 non-null object
expanded_urls                 2297 non-null object
rating_numerator              2356 non-null int64
rating_denominator            2356 non-null int64
name                          2356 non-null object
doggo                         2356 non-null object
floofer                       2356 non-null object
pupper                        2356 non-null object
puppo                         2356 non-null object
dtypes: float64(4), int64(3), ob

C:\Users\singhd5\AppData\Local\Continuum\Anaconda3\lib\site-packages\numpy\lib\function_base.py:3834: RuntimeWarning: Invalid value encountered in percentile
  RuntimeWarning)


None           745
a               55
Charlie         12
Lucy            11
Cooper          11
Oliver          11
Lola            10
Tucker          10
Penny           10
Winston          9
Bo               9
the              8
Sadie            8
Toby             7
Buddy            7
Daisy            7
Bailey           7
an               7
Milo             6
Jack             6
Bella            6
Scout            6
Oscar            6
Stanley          6
Dave             6
Leo              6
Jax              6
Rusty            6
Koda             6
Larry            5
              ... 
Augie            1
Dug              1
Mark             1
Zeek             1
Boston           1
Lili             1
Yoda             1
Pawnd            1
Obi              1
Zoe              1
Lucky            1
Kanu             1
Bloop            1
Emmie            1
Bluebert         1
Jessifer         1
Lilli            1
infuriating      1
Anna             1
Lupe             1
Gordon           1
Skittles    

In [102]:
archive.describe()

C:\Users\singhd5\AppData\Local\Continuum\Anaconda3\lib\site-packages\numpy\lib\function_base.py:3834: RuntimeWarning: Invalid value encountered in percentile
  RuntimeWarning)


tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
count  2.356000e+03           7.800000e+01         7.800000e+01   
mean   7.427716e+17           7.455079e+17         2.014171e+16   
std    6.856705e+16           7.582492e+16         1.252797e+17   
min    6.660209e+17           6.658147e+17         1.185634e+07   
25%    6.783989e+17                    NaN                  NaN   
50%    7.196279e+17                    NaN                  NaN   
75%    7.993373e+17                    NaN                  NaN   
max    8.924206e+17           8.862664e+17         8.405479e+17   

       retweeted_status_id  retweeted_status_user_id  rating_numerator  \
count         1.810000e+02              1.810000e+02       2356.000000   
mean          7.720400e+17              1.241698e+16         13.126486   
std           6.236928e+16              9.599254e+16         45.876648   
min           6.661041e+17              7.832140e+05          0.000000   
25%                    NaN                       NaN         10.000000   
50%                    NaN                       NaN         11.000000   
75%                    NaN                       NaN         12.000000   
max           8.874740e+17              7.874618e+17       1776.000000   

       rating_denominator  
count         2356.000000  
mean            10.455433  
std              6.745237  
min              0.000000  
25%             10.000000  
50%             10.000000  
75%             10.000000  
max            170.000000

In [110]:
image_predictions.head(20)


tweet_id                                          jpg_url  \
0   666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1   666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2   666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3   666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4   666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   
5   666050758794694657  https://pbs.twimg.com/media/CT5Jof1WUAEuVxN.jpg   
6   666051853826850816  https://pbs.twimg.com/media/CT5KoJ1WoAAJash.jpg   
7   666055525042405380  https://pbs.twimg.com/media/CT5N9tpXIAAifs1.jpg   
8   666057090499244032  https://pbs.twimg.com/media/CT5PY90WoAAQGLo.jpg   
9   666058600524156928  https://pbs.twimg.com/media/CT5Qw94XAAA_2dP.jpg   
10  666063827256086533  https://pbs.twimg.com/media/CT5Vg_wXIAAXfnj.jpg   
11  666071193221509120  https://pbs.twimg.com/media/CT5cN_3WEAAlOoZ.jpg   
12  666073100786774016  https://pbs.twimg.com/media/CT5d9DZXAAALcwe.jpg   
13  666082916733198337  https://pbs.twimg.com/media/CT5m4VGWEAAtKc8.jpg   
14  666094000022159362  https://pbs.twimg.com/media/CT5w9gUW4AAsBNN.jpg   
15  666099513787052032  https://pbs.twimg.com/media/CT51-JJUEAA6hV8.jpg   
16  666102155909144576  https://pbs.twimg.com/media/CT54YGiWUAEZnoK.jpg   
17  666104133288665088  https://pbs.twimg.com/media/CT56LSZWoAAlJj2.jpg   
18  666268910803644416  https://pbs.twimg.com/media/CT8QCd1WEAADXws.jpg   
19  666273097616637952  https://pbs.twimg.com/media/CT8T1mtUwAA3aqm.jpg   

    img_num                      p1   p1_conf p1_dog                  p2  \
0         1  Welsh_springer_spaniel  0.465074   True              collie   
1         1                 redbone  0.506826   True  miniature_pinscher   
2         1         German_shepherd  0.596461   True            malinois   
3         1     Rhodesian_ridgeback  0.408143   True             redbone   
4         1      miniature_pinscher  0.560311   True          Rottweiler   
5         1    Bernese_mountain_dog  0.651137   True    English_springer   
6         1              box_turtle  0.933012  False          mud_turtle   
7         1                    chow  0.692517   True     Tibetan_mastiff   
8         1           shopping_cart  0.962465  False     shopping_basket   
9         1        miniature_poodle  0.201493   True            komondor   
10        1        golden_retriever  0.775930   True     Tibetan_mastiff   
11        1           Gordon_setter  0.503672   True   Yorkshire_terrier   
12        1            Walker_hound  0.260857   True    English_foxhound   
13        1                     pug  0.489814   True        bull_mastiff   
14        1              bloodhound  0.195217   True     German_shepherd   
15        1                   Lhasa  0.582330   True            Shih-Tzu   
16        1          English_setter  0.298617   True        Newfoundland   
17        1                     hen  0.965932  False                cock   
18        1        desktop_computer  0.086502  False                desk   
19        1       Italian_greyhound  0.176053   True         toy_terrier   

     p2_conf p2_dog                           p3   p3_conf p3_dog  
0   0.156665   True            Shetland_sheepdog  0.061428   True  
1   0.074192   True          Rhodesian_ridgeback  0.072010   True  
2   0.138584   True                   bloodhound  0.116197   True  
3   0.360687   True           miniature_pinscher  0.222752   True  
4   0.243682   True                     Doberman  0.154629   True  
5   0.263788   True   Greater_Swiss_Mountain_dog  0.016199   True  
6   0.045885  False                     terrapin  0.017885  False  
7   0.058279   True                     fur_coat  0.054449  False  
8   0.014594  False             golden_retriever  0.007959   True  
9   0.192305   True  soft-coated_wheaten_terrier  0.082086   True  
10  0.093718   True           Labrador_retriever  0.072427   True  
11  0.174201   True      

In [109]:
image_predictions.tail(20)

tweet_id                                            jpg_url  \
2055  888202515573088257    https://pbs.twimg.com/media/DFDw2tyUQAAAFke.jpg   
2056  888554962724278272    https://pbs.twimg.com/media/DFTH_O-UQAACu20.jpg   
2057  888804989199671297    https://pbs.twimg.com/media/DFWra-3VYAA2piG.jpg   
2058  888917238123831296    https://pbs.twimg.com/media/DFYRgsOUQAARGhO.jpg   
2059  889278841981685760  https://pbs.twimg.com/ext_tw_video_thumb/88927...   
2060  889531135344209921    https://pbs.twimg.com/media/DFg_2PVW0AEHN3p.jpg   
2061  889638837579907072    https://pbs.twimg.com/media/DFihzFfXsAYGDPR.jpg   
2062  889665388333682689    https://pbs.twimg.com/media/DFi579UWsAAatzw.jpg   
2063  889880896479866881    https://pbs.twimg.com/media/DFl99B1WsAITKsg.jpg   
2064  890006608113172480    https://pbs.twimg.com/media/DFnwSY4WAAAMliS.jpg   
2065  890240255349198849    https://pbs.twimg.com/media/DFrEyVuW0AAO3t9.jpg   
2066  890609185150312448    https://pbs.twimg.com/media/DFwUU__XcAEpyXI.jpg   
2067  890729181411237888    https://pbs.twimg.com/media/DFyBahAVwAAhUTd.jpg   
2068  890971913173991426    https://pbs.twimg.com/media/DF1eOmZXUAALUcq.jpg   
2069  891087950875897856    https://pbs.twimg.com/media/DF3HwyEWsAABqE6.jpg   
2070  891327558926688256    https://pbs.twimg.com/media/DF6hr6BUMAAzZgT.jpg   
2071  891689557279858688    https://pbs.twimg.com/media/DF_q7IAWsAEuuN8.jpg   
2072  891815181378084864    https://pbs.twimg.com/media/DGBdLU1WsAANxJ9.jpg   
2073  892177421306343426    https://pbs.twimg.com/media/DGGmoV4XsAAUL6n.jpg   
2074  892420643555336193    https://pbs.twimg.com/media/DGKD1-bXoAAIAUK.jpg   

      img_num                        p1   p1_conf p1_dog                   p2  \
2055        2                  Pembroke  0.809197   True  Rhodesian_ridgeback   
2056        3            Siberian_husky  0.700377   True           Eskimo_dog   
2057        1          golden_retriever  0.469760   True   Labrador_retriever   
2058        1          golden_retriever  0.714719   True      Tibetan_mastiff   
2059        1                   whippet  0.626152   True               borzoi   
2060        1          golden_retriever  0.953442   True   Labrador_retriever   
2061        1            French_bulldog  0.991650   True                boxer   
2062        1                  Pembroke  0.966327   True             Cardigan   
2063        1            French_bulldog  0.377417   True   Labrador_retriever   
2064        1                   Samoyed  0.957979   True           Pomeranian   
2065        1                  Pembroke  0.511319   True             Cardigan   
2066        1             Irish_terrier  0.487574   True         Irish_setter   
2067        2                Pomeranian  0.566142   True           Eskimo_dog   
2068        1               Appenzeller  0.341703   True        Border_collie   
2069        1  Chesapeake_Bay_retriever  0.425595   True        Irish_terrier   
2070        2                    basset  0.555712   True     English_springer   
2071        1               paper_towel  0.170278  False   Labrador_retriever   
2072        1                 Chihuahua  0.716012   True             malamute   
2073        1                 Chihuahua  0.323581   True             Pekinese   
2074        1                    orange  0.097049  False                bagel   

       p2_conf p2_dog                           p3   p3_conf p3_dog  
2055  0.054950   True                       beagle  0.038915   True  
2056  0.166511   True                     malamute  0.111411   True  
2057  0.184172   True               English_setter  0.073482   True  
2058  0.120184   True           Labrador_retriever  0.105506   True  
2059  0.194742   True                       Saluki  0.027351   True  
2060  0.013834   True                      redbone  0.007958   True  
2061  0.002129   True    Staffordshire_bullterrier  0.001498   True  
2062  0.027356   True                      basenji  0.004633   True  
2063  0.151317   True   

In [111]:
image_predictions.info()
image_predictions.describe()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
tweet_id    2075 non-null int64
jpg_url     2075 non-null object
img_num     2075 non-null int64
p1          2075 non-null object
p1_conf     2075 non-null float64
p1_dog      2075 non-null bool
p2          2075 non-null object
p2_conf     2075 non-null float64
p2_dog      2075 non-null bool
p3          2075 non-null object
p3_conf     2075 non-null float64
p3_dog      2075 non-null bool
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB


tweet_id      img_num      p1_conf       p2_conf       p3_conf
count  2.075000e+03  2075.000000  2075.000000  2.075000e+03  2.075000e+03
mean   7.384514e+17     1.203855     0.594548  1.345886e-01  6.032417e-02
std    6.785203e+16     0.561875     0.271174  1.006657e-01  5.090593e-02
min    6.660209e+17     1.000000     0.044333  1.011300e-08  1.740170e-10
25%    6.764835e+17     1.000000     0.364412  5.388625e-02  1.622240e-02
50%    7.119988e+17     1.000000     0.588230  1.181810e-01  4.944380e-02
75%    7.932034e+17     1.000000     0.843855  1.955655e-01  9.180755e-02
max    8.924206e+17     4.000000     1.000000  4.880140e-01  2.734190e-01

In [112]:
image_predictions['jpg_url'].value_counts()

https://pbs.twimg.com/media/C2oRbOuWEAAbVSl.jpg                                            2
https://pbs.twimg.com/media/CVMOlMiWwAA4Yxl.jpg                                            2
https://pbs.twimg.com/media/C4bTH6nWMAAX_bJ.jpg                                            2
https://pbs.twimg.com/media/Ct2qO5PXEAE6eB0.jpg                                            2
https://pbs.twimg.com/media/CdHwZd0VIAA4792.jpg                                            2
https://pbs.twimg.com/media/CUN4Or5UAAAa5K4.jpg                                            2
https://pbs.twimg.com/media/Cveg1-NXgAASaaT.jpg                                            2
https://pbs.twimg.com/tweet_video_thumb/CeBym7oXEAEWbEg.jpg                                2
https://pbs.twimg.com/media/DA7iHL5U0AA1OQo.jpg                                            2
https://pbs.twimg.com/media/CVuQ2LeUsAAIe3s.jpg                                            2
https://pbs.twimg.com/media/CxqsX-8XUAAEvjD.jpg                       

In [114]:
image_predictions[image_predictions['jpg_url']== 'https://pbs.twimg.com/ext_tw_video_thumb/807106774843039744/pu/img/8XZg1xW35Xp2J6JW.jpg']

tweet_id                                            jpg_url  \
1641  807106840509214720  https://pbs.twimg.com/ext_tw_video_thumb/80710...   
1864  842892208864923648  https://pbs.twimg.com/ext_tw_video_thumb/80710...   

      img_num         p1  p1_conf p1_dog          p2   p2_conf p2_dog  \
1641        1  Chihuahua  0.50537   True  Pomeranian  0.120358   True   
1864        1  Chihuahua  0.50537   True  Pomeranian  0.120358   True   

               p3   p3_conf p3_dog  
1641  toy_terrier  0.077008   True  
1864  toy_terrier  0.077008   True

In [120]:
json_tweets.head(20)

created_at  favorite_count  retweet_count            tweet_id
0   2017-08-01 16:23:56           39014           8660  892420643555336193
1   2017-08-01 00:17:27           33403           6359  892177421306343426
2   2017-07-31 00:18:03           25189           4224  891815181378084864
3   2017-07-30 15:58:51           42368           8769  891689557279858688
4   2017-07-29 16:00:24           40523           9538  891327558926688256
5   2017-07-29 00:08:17           20327           3164  891087950875897856
6   2017-07-28 16:27:12           11917           2107  890971913173991426
7   2017-07-28 00:22:40           65932          19194  890729181411237888
8   2017-07-27 16:25:51           27911           4326  890609185150312448
9   2017-07-26 15:59:51           32118           7539  890240255349198849
10  2017-07-26 00:31:25           30814           7448  890006608113172480
11  2017-07-25 16:11:53           27932           5040  889880896479866881
12  2017-07-25 01:55:32           48332          10215  889665388333682689
13  2017-07-25 00:10:02           27341           4611  889638837579907072
14  2017-07-24 17:02:04           15174           2264  889531135344209921
15  2017-07-24 00:19:32           25430           5528  889278841981685760
16  2017-07-23 00:22:39           29252           4574  888917238123831296
17  2017-07-22 16:56:37           25740           4422  888804989199671297
18  2017-07-22 00:23:06           20036           3653  888554962724278272
19  2017-07-20 16:49:33           21886           3558  888078434458587136

In [121]:
json_tweets.tail(20)

created_at  favorite_count  retweet_count            tweet_id
2336  2015-11-16 00:55:59             299            143  666057090499244032
2337  2015-11-16 00:49:46             441            258  666055525042405380
2338  2015-11-16 00:35:11            1236            864  666051853826850816
2339  2015-11-16 00:30:50             133             59  666050758794694657
2340  2015-11-16 00:24:50             109             40  666049248165822465
2341  2015-11-16 00:04:52             302            141  666044226329800704
2342  2015-11-15 23:21:54             125             45  666033412701032449
2343  2015-11-15 23:05:30             131             47  666029285002620928
2344  2015-11-15 22:32:08            2521            519  666020888022790149
2345  2015-11-15 22:32:08            2521            519  888202515573088257
2346  2015-11-15 22:32:08            2521            519  873697596434513921
2347  2015-11-15 22:32:08            2521            519  869988702071779329
2348  2015-11-15 22:32:08            2521            519  866816280283807744
2349  2015-11-15 22:32:08            2521            519  861769973181624320
2350  2015-11-15 22:32:08            2521            519  845459076796616705
2351  2015-11-15 22:32:08            2521            519  842892208864923648
2352  2015-11-15 22:32:08            2521            519  837012587749474308
2353  2015-11-15 22:32:08            2521            519  827228250799742977
2354  2015-11-15 22:32:08            2521            519  802247111496568832
2355  2015-11-15 22:32:08            2521            519  775096608509886464

In [122]:
json_tweets.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 4 columns):
created_at        2356 non-null object
favorite_count    2356 non-null int64
retweet_count     2356 non-null int64
tweet_id          2356 non-null int64
dtypes: int64(3), object(1)
memory usage: 73.7+ KB


## Quality Issues Log

Some of the Completeness, Validity, Accuracy, Consistency issues are documented below

Issues with the data's content

Need to remove rows that have been retweeted, therby they are not original tweets

### A tweet that has been retweeted

The relevant field is retweet_count. This field provides the number of times this tweet was retweeted. Note that this number may vary over time, as additional people retweet the tweet.

### A tweet that is a retweet

Want to exclude any tweet that is a retweet. Two fields are significant. First, the retweeted_status contains the source tweet (i.e., the tweet that was retweeted). The present or absence of this field can be used to identify tweets that are retweets. Second, the retweet_count is the count of the retweets of the source tweet, not this tweet.

Therefore I will isolate all rows in the retweeted_status column that have a value and delete it from the archive dataframe. This will remove tweets that are a retweet from the dataframe.

archive dataframe:

- Some ratings numerator and denominator are incorrectly extracted
- Fix rating numerator that have decimals
- in_reply_to_status_id, in_reply_to_user_id, retweeted_status_id, retweeted_status_user_id should be integer instead of float 
- retweeted_status_timestamp, timestamp should be datetime instead of object (string)
- Names of dogs are miss labelled, mispelled or missing. Cross-reference text data with Names column
- 181 records have a retweeted_status_id, these will need to be exluded from the dataset 
- Remove extra characters after '&' in archive['text'].
- In several columns null objects are non-null 

image_predictions dataframe:

- Rename column names to more informative (pic_1, pic_2..)


## Tidiness Issue Log

Some of the structural issues in the involed dataset are captured below.

- Various stages of dogs in columns instead of rows in archive dataset
- keep only tweet_id and jpg_url columns in image_predictins dataframe
- join all three datasets


# Clean 

Cleaning our data is the third step in data wrangling. It is where we will fix the quality and tidiness issues that we identified in the assess step.


In [186]:
# Copy the dataframes 
df_clean = archive.copy()
images_clean = image_predictions.copy()
tweet_df_clean = json_tweets.copy()

### Define

Correct names of the dogs by replacing them with Null values 

### Code

In [187]:
# Missing Data
# replacing faulty names with None value or corrected Name
df_clean['name'].replace('the', 'None', inplace=True)
df_clean['name'].replace("light",'None', inplace=True)
df_clean['name'].replace("life",'None', inplace=True)
df_clean['name'].replace("an",'None', inplace=True)
df_clean['name'].replace("a",'None', inplace=True)
df_clean['name'].replace("by",'None', inplace=True)
df_clean['name'].replace("actually",'None', inplace=True)
df_clean['name'].replace("just",'None', inplace=True)
df_clean['name'].replace("getting",'None', inplace=True) 
df_clean['name'].replace("infuriating",'None', inplace=True) 
df_clean['name'].replace("old",'None', inplace=True) 
df_clean['name'].replace("all",'None', inplace=True) 
df_clean['name'].replace("this",'None', inplace=True) 
df_clean['name'].replace("very",'None', inplace=True) 
df_clean['name'].replace("mad",'None', inplace=True) 
df_clean['name'].replace("not",'None', inplace=True)
df_clean['name'].replace("one",'None', inplace=True)
df_clean['name'].replace("my",'None', inplace=True)
df_clean['name'].replace("O",'None', inplace=True)
df_clean['name'].replace("quite","None", inplace=True)
df_clean['name'].replace("such","None", inplace=True)

### Test

In [188]:
df_clean['name'].value_counts()

None         850
Charlie       12
Lucy          11
Cooper        11
Oliver        11
Tucker        10
Penny         10
Lola          10
Bo             9
Winston        9
Sadie          8
Toby           7
Buddy          7
Bailey         7
Daisy          7
Bella          6
Jax            6
Stanley        6
Milo           6
Rusty          6
Oscar          6
Scout          6
Leo            6
Jack           6
Koda           6
Dave           6
Bentley        5
Larry          5
Oakley         5
Sunny          5
            ... 
Smiley         1
Augie          1
Dug            1
Mark           1
Zeek           1
Lili           1
Lupe           1
Charl          1
Lilli          1
Harvey         1
Chadrick       1
Zoe            1
Lucky          1
Kanu           1
Bloop          1
Emmie          1
Bluebert       1
Jessifer       1
Obi            1
Al             1
Anna           1
Gordon         1
Skittles       1
Shnuggles      1
Rose           1
Dwight         1
Kial           1
Yukon         

In [ ]:
# Define -Convert colums in the asses log to desired datatypes
# Code
df_clean['in_reply_to_status_id'] = df_clean['in_reply_to_status_id'].astype(object)
df_clean['in_reply_to_user_id'] = df_clean['in_reply_to_user_id'].astype(object)
df_clean['retweeted_status_id'] = df_clean['retweeted_status_id'].astype(object)
df_clean['retweeted_status_user_id'] = df_clean['retweeted_status_user_id'].astype(object)
df_clean['retweeted_status_timestamp'] = pd.to_datetime(df_clean.retweeted_status_timestamp)


In [190]:
# Test
df_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2356 non-null int64
in_reply_to_status_id         78 non-null object
in_reply_to_user_id           78 non-null object
timestamp                     2356 non-null object
source                        2356 non-null object
text                          2356 non-null object
retweeted_status_id           181 non-null object
retweeted_status_user_id      181 non-null object
retweeted_status_timestamp    181 non-null datetime64[ns]
expanded_urls                 2297 non-null object
rating_numerator              2356 non-null int64
rating_denominator            2356 non-null int64
name                          2356 non-null object
doggo                         2356 non-null object
floofer                       2356 non-null object
pupper                        2356 non-null object
puppo                         2356 non-null object
dtypes: datetime64[ns](1), i

In [191]:
# Define - Convert NaN to None
# Code

df_clean['in_reply_to_status_id'].fillna(value='None',method=None, inplace = True)
df_clean['in_reply_to_user_id'].fillna(value='None',method=None, inplace = True)
df_clean['retweeted_status_id'].fillna(value='None',method=None, inplace = True)
df_clean['retweeted_status_user_id'].fillna(value='None',method=None, inplace = True)
df_clean['retweeted_status_timestamp'] .fillna(value='None',method=None, inplace = True)

In [192]:
# Test
df_clean.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2356 non-null int64
in_reply_to_status_id         2356 non-null object
in_reply_to_user_id           2356 non-null object
timestamp                     2356 non-null object
source                        2356 non-null object
text                          2356 non-null object
retweeted_status_id           2356 non-null object
retweeted_status_user_id      2356 non-null object
retweeted_status_timestamp    2356 non-null object
expanded_urls                 2297 non-null object
rating_numerator              2356 non-null int64
rating_denominator            2356 non-null int64
name                          2356 non-null object
doggo                         2356 non-null object
floofer                       2356 non-null object
pupper                        2356 non-null object
puppo                         2356 non-null object
dtypes: int64(3), object(14)


In [194]:
# Define - remove the 181 retweets from teh archive dataframe
# Code
df_clean[df_clean['retweeted_status_id'] != 'None'].info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 181 entries, 19 to 2260
Data columns (total 17 columns):
tweet_id                      181 non-null int64
in_reply_to_status_id         181 non-null object
in_reply_to_user_id           181 non-null object
timestamp                     181 non-null object
source                        181 non-null object
text                          181 non-null object
retweeted_status_id           181 non-null object
retweeted_status_user_id      181 non-null object
retweeted_status_timestamp    181 non-null object
expanded_urls                 180 non-null object
rating_numerator              181 non-null int64
rating_denominator            181 non-null int64
name                          181 non-null object
doggo                         181 non-null object
floofer                       181 non-null object
pupper                        181 non-null object
puppo                         181 non-null object
dtypes: int64(3), object(14)
memory usage: 25.

In [195]:
# code
df_clean.drop(df_clean[df_clean['retweeted_status_id'] != 'None'].index, inplace = True)

In [197]:
# test
df_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2175 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2175 non-null int64
in_reply_to_status_id         2175 non-null object
in_reply_to_user_id           2175 non-null object
timestamp                     2175 non-null object
source                        2175 non-null object
text                          2175 non-null object
retweeted_status_id           2175 non-null object
retweeted_status_user_id      2175 non-null object
retweeted_status_timestamp    2175 non-null object
expanded_urls                 2117 non-null object
rating_numerator              2175 non-null int64
rating_denominator            2175 non-null int64
name                          2175 non-null object
doggo                         2175 non-null object
floofer                       2175 non-null object
pupper                        2175 non-null object
puppo                         2175 non-null object
dtypes: int64(3), object(14)


In [238]:
# Define - Rename columns in image_pedictions file
# Code
images_clean= images_clean.rename(columns = {"p1":"pic_1","p2":"pic_2","p3":"pic_3","p4":"pic_4",
                                "p1_conf":"pic1_conf","p2_conf":"pic2_conf","p3_conf":"pic3_conf","p4_conf":"pic4_conf",
                              "p1_dog":"pic1_dog","p2_dog":"pic2_dog","p3_dog":"pic3_dog","p4_dog":"pic4_dog"})

In [204]:
images_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
tweet_id    2075 non-null int64
jpg_url     2075 non-null object
img_num     2075 non-null int64
p1          2075 non-null object
p1_conf     2075 non-null float64
p1_dog      2075 non-null bool
p2          2075 non-null object
p2_conf     2075 non-null float64
p2_dog      2075 non-null bool
p3          2075 non-null object
p3_conf     2075 non-null float64
p3_dog      2075 non-null bool
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB


In [205]:
# Define - Save the text where the rating numerator and denominators were incorrectly extracted
ratings_to_fix = ['After so many requests, this is Bretagne. She was the last surviving 9/11 search dog, and our second ever 14/10. RIP https://t.co/XAVDNDaVgQ', 
 'Happy 4/20 from the squad! 13/10 for all https://t.co/eV1diwds8a', 
 'This is Bluebert. He just saw that both #FinalFur match ups are split 50/50. Amazed af. 11/10 https://t.co/Kky1DPG4iq', 
 'This is Darrel. He just robbed a 7/11 and is in a high speed police chase. Was just spotted by the helicopter 10/10 https://t.co/7EsP8LmSp5',
 'This is an Albanian 3 1/2 legged  Episcopalian. Loves well-polished hardwood flooring. Penis on the collar. 9/10 https://t.co/d9NcXFKwLv']

In [209]:
# Code
# Loop through the list of ratings to fix and extract the second occurence of #/ to save as the rating numerator. As all the
# occurences of the actual ratings in the ratings to fix list have a denominator of 10, we will set that value for each 
#entry instead of extracting it.
for entry in ratings_to_fix:
    mask = df_clean.text == entry
    column_name1 = 'rating_numerator'
    column_name2 = 'rating_denominator'
    df_clean.loc[mask, column_name1] = re.findall(r"\d+\.?\d*\/\d+\.?\d*\D+(\d+\.?\d*)\/\d+\.?\d*", entry)
    df_clean.loc[mask, column_name2] = 10

In [210]:
# Test
df_clean[df_clean.text.isin(ratings_to_fix)]

tweet_id in_reply_to_status_id in_reply_to_user_id  \
1068  740373189193256964                  None                None   
1165  722974582966214656                  None                None   
1202  716439118184652801                  None                None   
1662  682962037429899265                  None                None   
2335  666287406224695296                  None                None   

                      timestamp  \
1068  2016-06-08 02:41:38 +0000   
1165  2016-04-21 02:25:47 +0000   
1202  2016-04-03 01:36:11 +0000   
1662  2016-01-01 16:30:13 +0000   
2335  2015-11-16 16:11:11 +0000   

                                                 source  \
1068  <a href="http://twitter.com/download/iphone" r...   
1165  <a href="http://twitter.com/download/iphone" r...   
1202  <a href="http://twitter.com/download/iphone" r...   
1662  <a href="http://twitter.com/download/iphone" r...   
2335  <a href="http://twitter.com/download/iphone" r...   

                                                   text retweeted_status_id  \
1068  After so many requests, this is Bretagne. She ...                None   
1165  Happy 4/20 from the squad! 13/10 for all https...                None   
1202  This is Bluebert. He just saw that both #Final...                None   
1662  This is Darrel. He just robbed a 7/11 and is i...                None   
2335  This is an Albanian 3 1/2 legged  Episcopalian...                None   

     retweeted_status_user_id retweeted_status_timestamp  \
1068                     None                       None   
1165                     None                       None   
1202                     None                       None   
1662                     None                       None   
2335                     None                       None   

                                          expanded_urls rating_numerator  \
1068  https://twitter.com/dog_rates/status/740373189...               14   
1165  https://twitter.com/dog_rates/status/722974582...               13   
1202  https://twitter.com/dog_rates/status/716439118...               11   
1662  https://twitter.com/dog_rates/status/682962037...               10   
2335  https://twitter.com/dog_rates/status/666287406...                9   

      rating_denominator      name doggo floofer pupper puppo  
1068                  10      None  None    None   None  None  
1165                  10      None  None    None   None  None  
1202                  10  Bluebert  None    None   None  None  
1662                  10    Darrel  None    None   None  None  
2335                  10      None  None    None   None  None

In [211]:
df_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2175 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2175 non-null int64
in_reply_to_status_id         2175 non-null object
in_reply_to_user_id           2175 non-null object
timestamp                     2175 non-null object
source                        2175 non-null object
text                          2175 non-null object
retweeted_status_id           2175 non-null object
retweeted_status_user_id      2175 non-null object
retweeted_status_timestamp    2175 non-null object
expanded_urls                 2117 non-null object
rating_numerator              2175 non-null object
rating_denominator            2175 non-null int64
name                          2175 non-null object
doggo                         2175 non-null object
floofer                       2175 non-null object
pupper                        2175 non-null object
puppo                         2175 non-null object
dtypes: int64(2), object(15)

In [215]:
df_clean['rating_numerator'] = df_clean['rating_numerator'].astype(float)
df_clean['rating_denominator'] = df_clean['rating_denominator'].astype(float)

In [216]:
df_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2175 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2175 non-null int64
in_reply_to_status_id         2175 non-null object
in_reply_to_user_id           2175 non-null object
timestamp                     2175 non-null object
source                        2175 non-null object
text                          2175 non-null object
retweeted_status_id           2175 non-null object
retweeted_status_user_id      2175 non-null object
retweeted_status_timestamp    2175 non-null object
expanded_urls                 2117 non-null object
rating_numerator              2175 non-null float64
rating_denominator            2175 non-null float64
name                          2175 non-null object
doggo                         2175 non-null object
floofer                       2175 non-null object
pupper                        2175 non-null object
puppo                         2175 non-null object
dtypes: float64(2), int64

In [217]:
# Define - Fix rating numerator that have decimals.
# Code
# View tweets with decimals in rating in 'text' column
df_clean[df_clean.text.str.contains(r"(\d+\.\d*\/\d+)")]

C:\Users\singhd5\AppData\Local\Continuum\Anaconda3\lib\site-packages\ipykernel\__main__.py:4: UserWarning: This pattern has match groups. To actually get the groups, use str.extract.


tweet_id in_reply_to_status_id in_reply_to_user_id  \
45    883482846933004288                  None                None   
695   786709082849828864                  None                None   
763   778027034220126208                  None                None   
1689  681340665377193984           6.81339e+17         4.19698e+09   
1712  680494726643068929                  None                None   

                      timestamp  \
45    2017-07-08 00:28:19 +0000   
695   2016-10-13 23:23:56 +0000   
763   2016-09-20 00:24:34 +0000   
1689  2015-12-28 05:07:27 +0000   
1712  2015-12-25 21:06:00 +0000   

                                                 source  \
45    <a href="http://twitter.com/download/iphone" r...   
695   <a href="http://twitter.com/download/iphone" r...   
763   <a href="http://twitter.com/download/iphone" r...   
1689  <a href="http://twitter.com/download/iphone" r...   
1712  <a href="http://twitter.com/download/iphone" r...   

                                                   text retweeted_status_id  \
45    This is Bella. She hopes her smile made you sm...                None   
695   This is Logan, the Chow who lived. He solemnly...                None   
763   This is Sophie. She's a Jubilant Bush Pupper. ...                None   
1689  I've been told there's a slight possibility he...                None   
1712  Here we have uncovered an entire battalion of ...                None   

     retweeted_status_user_id retweeted_status_timestamp  \
45                       None                       None   
695                      None                       None   
763                      None                       None   
1689                     None                       None   
1712                     None                       None   

                                          expanded_urls  rating_numerator  \
45    https://twitter.com/dog_rates/status/883482846...               5.0   
695   https://twitter.com/dog_rates/status/786709082...              75.0   
763   https://twitter.com/dog_rates/status/778027034...              27.0   
1689                                                NaN               5.0   
1712  https://twitter.com/dog_rates/status/680494726...              26.0   

      rating_denominator    name doggo floofer  pupper puppo  
45                  10.0   Bella  None    None    None  None  
695                 10.0   Logan  None    None    None  None  
763                 10.0  Sophie  None    None  pupper  None  
1689                10.0    None  None    None    None  None  
1712                10.0    None  None    None    None  None

In [219]:
# Set correct numerators for specific tweets
df_clean.loc[(df_clean['tweet_id'] == 883482846933004288) & (df_clean['rating_numerator'] == 5), ['rating_numerator']] = 13.5
df_clean.loc[(df_clean['tweet_id'] == 786709082849828864) & (df_clean['rating_numerator'] == 75), ['rating_numerator']] = 9.75
df_clean.loc[(df_clean['tweet_id'] == 778027034220126208) & (df_clean['rating_numerator'] == 27), ['rating_numerator']] = 11.27
df_clean.loc[(df_clean['tweet_id'] == 680494726643068929) & (df_clean['rating_numerator'] == 26), ['rating_numerator']] = 11.26

In [220]:
# Test
df_clean[df_clean.text.str.contains(r"(\d+\.\d*\/\d+)")]

C:\Users\singhd5\AppData\Local\Continuum\Anaconda3\lib\site-packages\ipykernel\__main__.py:2: UserWarning: This pattern has match groups. To actually get the groups, use str.extract.
  from ipykernel import kernelapp as app


tweet_id in_reply_to_status_id in_reply_to_user_id  \
45    883482846933004288                  None                None   
695   786709082849828864                  None                None   
763   778027034220126208                  None                None   
1689  681340665377193984           6.81339e+17         4.19698e+09   
1712  680494726643068929                  None                None   

                      timestamp  \
45    2017-07-08 00:28:19 +0000   
695   2016-10-13 23:23:56 +0000   
763   2016-09-20 00:24:34 +0000   
1689  2015-12-28 05:07:27 +0000   
1712  2015-12-25 21:06:00 +0000   

                                                 source  \
45    <a href="http://twitter.com/download/iphone" r...   
695   <a href="http://twitter.com/download/iphone" r...   
763   <a href="http://twitter.com/download/iphone" r...   
1689  <a href="http://twitter.com/download/iphone" r...   
1712  <a href="http://twitter.com/download/iphone" r...   

                                                   text retweeted_status_id  \
45    This is Bella. She hopes her smile made you sm...                None   
695   This is Logan, the Chow who lived. He solemnly...                None   
763   This is Sophie. She's a Jubilant Bush Pupper. ...                None   
1689  I've been told there's a slight possibility he...                None   
1712  Here we have uncovered an entire battalion of ...                None   

     retweeted_status_user_id retweeted_status_timestamp  \
45                       None                       None   
695                      None                       None   
763                      None                       None   
1689                     None                       None   
1712                     None                       None   

                                          expanded_urls  rating_numerator  \
45    https://twitter.com/dog_rates/status/883482846...             13.50   
695   https://twitter.com/dog_rates/status/786709082...              9.75   
763   https://twitter.com/dog_rates/status/778027034...             11.27   
1689                                                NaN              5.00   
1712  https://twitter.com/dog_rates/status/680494726...             11.26   

      rating_denominator    name doggo floofer  pupper puppo  
45                  10.0   Bella  None    None    None  None  
695                 10.0   Logan  None    None    None  None  
763                 10.0  Sophie  None    None  pupper  None  
1689                10.0    None  None    None    None  None  
1712                10.0    None  None    None    None  None

In [224]:
tweet_df_clean.drop_duplicates(subset=['tweet_id'])

created_at  favorite_count  retweet_count            tweet_id
0     2017-08-01 16:23:56           39014           8660  892420643555336193
1     2017-08-01 00:17:27           33403           6359  892177421306343426
2     2017-07-31 00:18:03           25189           4224  891815181378084864
3     2017-07-30 15:58:51           42368           8769  891689557279858688
4     2017-07-29 16:00:24           40523           9538  891327558926688256
5     2017-07-29 00:08:17           20327           3164  891087950875897856
6     2017-07-28 16:27:12           11917           2107  890971913173991426
7     2017-07-28 00:22:40           65932          19194  890729181411237888
8     2017-07-27 16:25:51           27911           4326  890609185150312448
9     2017-07-26 15:59:51           32118           7539  890240255349198849
10    2017-07-26 00:31:25           30814           7448  890006608113172480
11    2017-07-25 16:11:53           27932           5040  889880896479866881
12    2017-07-25 01:55:32           48332          10215  889665388333682689
13    2017-07-25 00:10:02           27341           4611  889638837579907072
14    2017-07-24 17:02:04           15174           2264  889531135344209921
15    2017-07-24 00:19:32           25430           5528  889278841981685760
16    2017-07-23 00:22:39           29252           4574  888917238123831296
17    2017-07-22 16:56:37           25740           4422  888804989199671297
18    2017-07-22 00:23:06           20036           3653  888554962724278272
19    2017-07-20 16:49:33           21886           3558  888078434458587136
20    2017-07-19 16:06:48           30358           5474  887705289381826560
21    2017-07-19 03:39:09           46474          11839  887517139158093824
22    2017-07-19 00:47:34           69510          18507  887473957103951883
23    2017-07-18 16:08:03           33857          10554  887343217045368832
24    2017-07-18 00:07:08           30722           6052  887101392804085760
25    2017-07-17 16:17:36           35380           7885  886983233522544640
26    2017-07-16 23:58:41           12147           3351  886736880519319552
27    2017-07-16 20:14:00           22560           4539  886680336477933568
28    2017-07-15 23:25:31           21241           3234  886366144734445568
29    2017-07-15 16:51:35             117              4  886267009285017600
...                   ...             ...            ...                 ...
2326  2015-11-16 14:57:41             105             36  666268910803644416
2327  2015-11-16 04:02:55           14496           6712  666104133288665088
2328  2015-11-16 03:55:04              80             13  666102155909144576
2329  2015-11-16 03:44:34             158             71  666099513787052032
2330  2015-11-16 03:22:39             166             76  666094000022159362
2331  2015-11-16 02:38:37             119             45  666082916733198337
2332  2015-11-16 01:59:36             326            172  666073100786774016
2333  2015-11-16 01:52:02             149             63  666071193221509120
2334  2015-11-16 01:22:45             482            224  666063827256086533
2335  2015-11-16 01:01:59             114             58  666058600524156928
2336  2015-11-16 00:55:59             299            143  666057090499244032
2337  2015-11-16 00:49:46             441            258  666055525042405380
2338  2015-11-16 00:35:11            1236            864  666051853826850816
2339  2015-11-16 00:30:50             133             59  666050758794694657
2340  2015-11-16 00:24:50             109             40  666049248165822465
2341  2015-11-16 00:04:52             302            141  666044226329800704
2342  2015-11-15 23:21:54             125             45  666033412701032449
2343  2015-11-15 23:05:30             131             47  666029285002620928
2344  2015-11-15 22:32:08            2521            519  666020888022790149
2345  2015-11-15 22:32:08            2521            519  888202515573088257
2346  2015-

In [225]:
# Define - Remove extra characters after '&' in Text of the tweet
# Code
df_clean['text'] = df_clean['text'].str.replace('&amp;', '&')

In [226]:
# Test
df_clean[df_clean.text.str.contains('&amp;')]

Empty DataFrame
Columns: [tweet_id, in_reply_to_status_id, in_reply_to_user_id, timestamp, source, text, retweeted_status_id, retweeted_status_user_id, retweeted_status_timestamp, expanded_urls, rating_numerator, rating_denominator, name, doggo, floofer, pupper, puppo]
Index: []

In [227]:
# Define - Create dog stage variable and remove individual dog stage columns
# Code
df_clean['dog_stage'] = df_clean['text'].str.extract('(puppo|pupper|floofer|doggo)', expand=True)

In [229]:
# remove the columns from the dataframe
columns = ['doggo', 'floofer', 'pupper', 'puppo']
df_clean = df_clean.drop(columns, axis=1)

In [230]:
# Test
df_clean.head(20)

tweet_id in_reply_to_status_id in_reply_to_user_id  \
0   892420643555336193                  None                None   
1   892177421306343426                  None                None   
2   891815181378084864                  None                None   
3   891689557279858688                  None                None   
4   891327558926688256                  None                None   
5   891087950875897856                  None                None   
6   890971913173991426                  None                None   
7   890729181411237888                  None                None   
8   890609185150312448                  None                None   
9   890240255349198849                  None                None   
10  890006608113172480                  None                None   
11  889880896479866881                  None                None   
12  889665388333682689                  None                None   
13  889638837579907072                  None                None   
14  889531135344209921                  None                None   
15  889278841981685760                  None                None   
16  888917238123831296                  None                None   
17  888804989199671297                  None                None   
18  888554962724278272                  None                None   
20  888078434458587136                  None                None   

                    timestamp  \
0   2017-08-01 16:23:56 +0000   
1   2017-08-01 00:17:27 +0000   
2   2017-07-31 00:18:03 +0000   
3   2017-07-30 15:58:51 +0000   
4   2017-07-29 16:00:24 +0000   
5   2017-07-29 00:08:17 +0000   
6   2017-07-28 16:27:12 +0000   
7   2017-07-28 00:22:40 +0000   
8   2017-07-27 16:25:51 +0000   
9   2017-07-26 15:59:51 +0000   
10  2017-07-26 00:31:25 +0000   
11  2017-07-25 16:11:53 +0000   
12  2017-07-25 01:55:32 +0000   
13  2017-07-25 00:10:02 +0000   
14  2017-07-24 17:02:04 +0000   
15  2017-07-24 00:19:32 +0000   
16  2017-07-23 00:22:39 +0000   
17  2017-07-22 16:56:37 +0000   
18  2017-07-22 00:23:06 +0000   
20  2017-07-20 16:49:33 +0000   

                                               source  \
0   <a href="http://twitter.com/download/iphone" r...   
1   <a href="http://twitter.com/download/iphone" r...   
2   <a href="http://twitter.com/download/iphone" r...   
3   <a href="http://twitter.com/download/iphone" r...   
4   <a href="http://twitter.com/download/iphone" r...   
5   <a href="http://twitter.com/download/iphone" r...   
6   <a href="http://twitter.com/download/iphone" r...   
7   <a href="http://twitter.com/download/iphone" r...   
8   <a href="http://twitter.com/download/iphone" r...   
9   <a href="http://twitter.com/download/iphone" r...   
10  <a href="http://twitter.com/download/iphone" r...   
11  <a href="http://twitter.com/download/iphone" r...   
12  <a href="http://twitter.com/download/iphone" r...   
13  <a href="http://twitter.com/download/iphone" r...   
14  <a href="http://twitter.com/download/iphone" r...   
15  <a href="http://twitter.com/download/iphone" r...   
16  <a href="http://twitter.com/download/iphone" r...   
17  <a href="http://twitter.com/download/iphone" r...   
18  <a href="http://twitter.com/download/iphone" r...   
20  <a href="http://twitter.com/download/iphone" r...   

                                                 text retweeted_status_id  \
0   This is Phineas. He's a mystical boy. Only eve...                None   
1   This is Tilly. She's just checking pup on you....                None   
2   This is Archie. He is a rare Norwegian Pouncin...                None   
3   This is Darla. She commenced a snooze mid meal...                None   
4   This is Franklin. He would like you to stop ca...                None   
5   Here we have a majestic great white breaching ...                None   
6   Meet Jax. He enjoys ice cream so much he gets ...                None   
7   When you watch your owner call another dog a g...                None   


In [234]:
# make all missing values to NaN
df_clean = df_clean.replace('None','NaN')

In [235]:
# Test
df_clean[df_clean['dog_stage'].isnull()]

tweet_id in_reply_to_status_id in_reply_to_user_id  \
0     892420643555336193                   NaN                 NaN   
1     892177421306343426                   NaN                 NaN   
2     891815181378084864                   NaN                 NaN   
3     891689557279858688                   NaN                 NaN   
4     891327558926688256                   NaN                 NaN   
5     891087950875897856                   NaN                 NaN   
6     890971913173991426                   NaN                 NaN   
7     890729181411237888                   NaN                 NaN   
8     890609185150312448                   NaN                 NaN   
10    890006608113172480                   NaN                 NaN   
11    889880896479866881                   NaN                 NaN   
13    889638837579907072                   NaN                 NaN   
15    889278841981685760                   NaN                 NaN   
16    888917238123831296                   NaN                 NaN   
17    888804989199671297                   NaN                 NaN   
18    888554962724278272                   NaN                 NaN   
20    888078434458587136                   NaN                 NaN   
21    887705289381826560                   NaN                 NaN   
22    887517139158093824                   NaN                 NaN   
23    887473957103951883                   NaN                 NaN   
24    887343217045368832                   NaN                 NaN   
25    887101392804085760                   NaN                 NaN   
26    886983233522544640                   NaN                 NaN   
27    886736880519319552                   NaN                 NaN   
28    886680336477933568                   NaN                 NaN   
30    886267009285017600           8.86266e+17         2.28118e+09   
31    886258384151887873                   NaN                 NaN   
33    885984800019947520                   NaN                 NaN   
34    885528943205470208                   NaN                 NaN   
35    885518971528720385                   NaN                 NaN   
...                  ...                   ...                 ...   
2326  666411507551481857                   NaN                 NaN   
2327  666407126856765440                   NaN                 NaN   
2328  666396247373291520                   NaN                 NaN   
2329  666373753744588802                   NaN                 NaN   
2330  666362758909284353                   NaN                 NaN   
2331  666353288456101888                   NaN                 NaN   
2332  666345417576210432                   NaN                 NaN   
2333  666337882303524864                   NaN                 NaN   
2334  666293911632134144                   NaN                 NaN   
2335  666287406224695296                   NaN                 NaN   
2336  666273097616637952                   NaN                 NaN   
2337  666268910803644416                   NaN                 NaN   
2338  666104133288665088                   NaN                 NaN   
2339  666102155909144576                   NaN                 NaN   
2340  666099513787052032                   NaN                 NaN   
2341  666094000022159362                   NaN                 NaN   
2342  666082916733198337                   NaN                 NaN   
2343  666073100786774016                   NaN                 NaN   
2344  666071193221509120                   NaN                 NaN   
2345  666063827256086533                   NaN                 NaN   
2346  666058600524156928                   NaN                 NaN   
2347  666057090499244032                   NaN                 NaN   
2348  666055525042405380                   NaN                 NaN   
2349  666051853826850816                   NaN                 NaN   
2350  666050758794694657                   NaN                 NaN   
2351  666049248165822465  

In [239]:
# Merge the three file into one master dataframe
# Code
df_clean = pd.merge(left=df_clean, right=images_clean, left_on='tweet_id', right_on='tweet_id', how='inner')


In [243]:
df_clean = pd.merge(left=df_clean, right=tweet_df_clean, left_on='tweet_id', right_on='tweet_id', how='inner')

In [246]:
df_clean.head(30)

tweet_id in_reply_to_status_id in_reply_to_user_id  \
0   892420643555336193                   NaN                 NaN   
1   892177421306343426                   NaN                 NaN   
2   891815181378084864                   NaN                 NaN   
3   891689557279858688                   NaN                 NaN   
4   891327558926688256                   NaN                 NaN   
5   891087950875897856                   NaN                 NaN   
6   890971913173991426                   NaN                 NaN   
7   890729181411237888                   NaN                 NaN   
8   890609185150312448                   NaN                 NaN   
9   890240255349198849                   NaN                 NaN   
10  890006608113172480                   NaN                 NaN   
11  889880896479866881                   NaN                 NaN   
12  889665388333682689                   NaN                 NaN   
13  889638837579907072                   NaN                 NaN   
14  889531135344209921                   NaN                 NaN   
15  889278841981685760                   NaN                 NaN   
16  888917238123831296                   NaN                 NaN   
17  888804989199671297                   NaN                 NaN   
18  888554962724278272                   NaN                 NaN   
19  888078434458587136                   NaN                 NaN   
20  887705289381826560                   NaN                 NaN   
21  887517139158093824                   NaN                 NaN   
22  887473957103951883                   NaN                 NaN   
23  887343217045368832                   NaN                 NaN   
24  887101392804085760                   NaN                 NaN   
25  886983233522544640                   NaN                 NaN   
26  886736880519319552                   NaN                 NaN   
27  886680336477933568                   NaN                 NaN   
28  886366144734445568                   NaN                 NaN   
29  886258384151887873                   NaN                 NaN   

                    timestamp  \
0   2017-08-01 16:23:56 +0000   
1   2017-08-01 00:17:27 +0000   
2   2017-07-31 00:18:03 +0000   
3   2017-07-30 15:58:51 +0000   
4   2017-07-29 16:00:24 +0000   
5   2017-07-29 00:08:17 +0000   
6   2017-07-28 16:27:12 +0000   
7   2017-07-28 00:22:40 +0000   
8   2017-07-27 16:25:51 +0000   
9   2017-07-26 15:59:51 +0000   
10  2017-07-26 00:31:25 +0000   
11  2017-07-25 16:11:53 +0000   
12  2017-07-25 01:55:32 +0000   
13  2017-07-25 00:10:02 +0000   
14  2017-07-24 17:02:04 +0000   
15  2017-07-24 00:19:32 +0000   
16  2017-07-23 00:22:39 +0000   
17  2017-07-22 16:56:37 +0000   
18  2017-07-22 00:23:06 +0000   
19  2017-07-20 16:49:33 +0000   
20  2017-07-19 16:06:48 +0000   
21  2017-07-19 03:39:09 +0000   
22  2017-07-19 00:47:34 +0000   
23  2017-07-18 16:08:03 +0000   
24  2017-07-18 00:07:08 +0000   
25  2017-07-17 16:17:36 +0000   
26  2017-07-16 23:58:41 +0000   
27  2017-07-16 20:14:00 +0000   
28  2017-07-15 23:25:31 +0000   
29  2017-07-15 16:17:19 +0000   

                                               source  \
0   <a href="http://twitter.com/download/iphone" r...   
1   <a href="http://twitter.com/download/iphone" r...   
2   <a href="http://twitter.com/download/iphone" r...   
3   <a href="http://twitter.com/download/iphone" r...   
4   <a href="http://twitter.com/download/iphone" r...   
5   <a href="http://twitter.com/download/iphone" r...   
6   <a href="http://twitter.com/download/iphone" r...   
7   <a href="http://twitter.com/download/iphone" r...   
8   <a href="http://twitter.com/download/iphone" r...   
9   <a href="http://twitter.com/download/iphone" r...   
10  <a href="http://twitter.com/download/iphone" r...   
11  <a href="http://twitter.com/download/iphone" r...   
12  <a href="http://twitter.com/download/iphone" r...   
13  <a href="http://twitter.com/download/iphone" r...   
14  <a href="http://twitte

In [247]:
columns=['p1','p1_conf','p1_dog','p2','p2_conf','p2_dog','p3','p3_conf','p3_dog']
df_clean = df_clean.drop(columns, axis=1)

In [248]:
df_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1994 entries, 0 to 1993
Data columns (total 30 columns):
tweet_id                      1994 non-null int64
in_reply_to_status_id         1994 non-null object
in_reply_to_user_id           1994 non-null object
timestamp                     1994 non-null object
source                        1994 non-null object
text                          1994 non-null object
retweeted_status_id           1994 non-null object
retweeted_status_user_id      1994 non-null object
retweeted_status_timestamp    1994 non-null object
expanded_urls                 1994 non-null object
rating_numerator              1994 non-null float64
rating_denominator            1994 non-null float64
name                          1994 non-null object
dog_stage                     326 non-null object
jpg_url_x                     1994 non-null object
img_num_x                     1994 non-null int64
jpg_url_y                     1994 non-null object
img_num_y                  

In [250]:
# Remove duplicates if any
df_clean = df_clean.drop_duplicates()

In [251]:
df_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1994 entries, 0 to 1993
Data columns (total 30 columns):
tweet_id                      1994 non-null int64
in_reply_to_status_id         1994 non-null object
in_reply_to_user_id           1994 non-null object
timestamp                     1994 non-null object
source                        1994 non-null object
text                          1994 non-null object
retweeted_status_id           1994 non-null object
retweeted_status_user_id      1994 non-null object
retweeted_status_timestamp    1994 non-null object
expanded_urls                 1994 non-null object
rating_numerator              1994 non-null float64
rating_denominator            1994 non-null float64
name                          1994 non-null object
dog_stage                     326 non-null object
jpg_url_x                     1994 non-null object
img_num_x                     1994 non-null int64
jpg_url_y                     1994 non-null object
img_num_y                  

In [252]:
# Remove any tweet with no URLs
# Code
df_clean = df_clean.dropna(subset=['expanded_urls'])

In [253]:
# Test
sum(df_clean['expanded_urls'].isnull())

0

In [262]:
# Remove and Rename the colums that are not requied
columns=['jpg_url_y','img_num_y','retweeted_status_id','retweeted_status_user_id','retweeted_status_timestamp']
df_clean = df_clean.drop(columns, axis=1)

In [267]:
df_clean = df_clean.rename(columns={"jpg_url_x":"jpg_url","img_num_x":"img_num"})

In [271]:
df_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1994 entries, 0 to 1993
Data columns (total 25 columns):
tweet_id                 1994 non-null int64
in_reply_to_status_id    1994 non-null object
in_reply_to_user_id      1994 non-null object
timestamp                1994 non-null object
source                   1994 non-null object
text                     1994 non-null object
expanded_urls            1994 non-null object
rating_numerator         1994 non-null float64
rating_denominator       1994 non-null float64
name                     1994 non-null object
dog_stage                326 non-null object
jpg_url                  1994 non-null object
img_num                  1994 non-null int64
pic_1                    1994 non-null object
pic1_conf                1994 non-null float64
pic1_dog                 1994 non-null bool
pic_2                    1994 non-null object
pic2_conf                1994 non-null float64
pic2_dog                 1994 non-null bool
pic_3                 

In [269]:
df_clean

tweet_id in_reply_to_status_id in_reply_to_user_id  \
0     892420643555336193                   NaN                 NaN   
1     892177421306343426                   NaN                 NaN   
2     891815181378084864                   NaN                 NaN   
3     891689557279858688                   NaN                 NaN   
4     891327558926688256                   NaN                 NaN   
5     891087950875897856                   NaN                 NaN   
6     890971913173991426                   NaN                 NaN   
7     890729181411237888                   NaN                 NaN   
8     890609185150312448                   NaN                 NaN   
9     890240255349198849                   NaN                 NaN   
10    890006608113172480                   NaN                 NaN   
11    889880896479866881                   NaN                 NaN   
12    889665388333682689                   NaN                 NaN   
13    889638837579907072                   NaN                 NaN   
14    889531135344209921                   NaN                 NaN   
15    889278841981685760                   NaN                 NaN   
16    888917238123831296                   NaN                 NaN   
17    888804989199671297                   NaN                 NaN   
18    888554962724278272                   NaN                 NaN   
19    888078434458587136                   NaN                 NaN   
20    887705289381826560                   NaN                 NaN   
21    887517139158093824                   NaN                 NaN   
22    887473957103951883                   NaN                 NaN   
23    887343217045368832                   NaN                 NaN   
24    887101392804085760                   NaN                 NaN   
25    886983233522544640                   NaN                 NaN   
26    886736880519319552                   NaN                 NaN   
27    886680336477933568                   NaN                 NaN   
28    886366144734445568                   NaN                 NaN   
29    886258384151887873                   NaN                 NaN   
...                  ...                   ...                 ...   
1964  666411507551481857                   NaN                 NaN   
1965  666407126856765440                   NaN                 NaN   
1966  666396247373291520                   NaN                 NaN   
1967  666373753744588802                   NaN                 NaN   
1968  666362758909284353                   NaN                 NaN   
1969  666353288456101888                   NaN                 NaN   
1970  666345417576210432                   NaN                 NaN   
1971  666337882303524864                   NaN                 NaN   
1972  666293911632134144                   NaN                 NaN   
1973  666287406224695296                   NaN                 NaN   
1974  666273097616637952                   NaN                 NaN   
1975  666268910803644416                   NaN                 NaN   
1976  666104133288665088                   NaN                 NaN   
1977  666102155909144576                   NaN                 NaN   
1978  666099513787052032                   NaN                 NaN   
1979  666094000022159362                   NaN                 NaN   
1980  666082916733198337                   NaN                 NaN   
1981  666073100786774016                   NaN                 NaN   
1982  666071193221509120                   NaN                 NaN   
1983  666063827256086533                   NaN                 NaN   
1984  666058600524156928                   NaN                 NaN   
1985  666057090499244032                   NaN                 NaN   
1986  666055525042405380                   NaN                 NaN   
1987  666051853826850816                   NaN                 NaN   
1988  666050758794694657                   NaN                 NaN   
1989  666049248165822465  

# Store



In [270]:
# Save clean DataFrame to csv file
df_clean.to_csv('twitter_archive_master.csv')